networkit/cpp/generators/ClusteredRandomGraphGenerator.cpp

```diff
@@ -3,55 +3,122 @@
  *      Author: cls
+ *              Eugenio Angriman <angrimae@hu-berlin.de>
  */
 
+// networkit-format
+
+#include <random>
+#include <stdexcept>
+#include <unordered_set>
+
 #include <networkit/auxiliary/Random.hpp>
 #include <networkit/generators/ClusteredRandomGraphGenerator.hpp>
-#include <networkit/structures/Partition.hpp>
+#include <networkit/graph/GraphBuilder.hpp>
 
 namespace NetworKit {
 
-ClusteredRandomGraphGenerator::ClusteredRandomGraphGenerator(count n, count k, double pin, double pout) : n(n), k(k), pin(pin), pout(pout) {
+ClusteredRandomGraphGenerator::ClusteredRandomGraphGenerator(count n, count k, double pIntra,
+                                                             double pInter)
+    : n(n), k(k), pIntra(pIntra), pInter(pInter), zeta(n) {
+
+    if (k == 0)
+        throw std::runtime_error("Error: k must be positive");
+
+    const auto isProbability = [](double p) noexcept -> bool { return (p >= 0) && (p <= 1); };
+
+    if (!isProbability(pIntra))
+        throw std::runtime_error("Error: pIntra must be in [0, 1]");
+
+    if (!isProbability(pInter))
+        throw std::runtime_error("Error: pInter must be in [0, 1]");
+
+    if (pIntra < pInter)
+        WARN("Intra-cluster probability is lower than inter-cluster probability.");
+
+    zeta.setUpperBound(k);
 }
 
 Graph ClusteredRandomGraphGenerator::generate() {
-    assert(pin >= pout);
+    GraphBuilder graphBuilder(n);
 
-    Graph G(n);
-    // assign nodes evenly to clusters
-    Partition zeta(n);
-    zeta.setUpperBound(k);
-    G.forNodes([&](node v){
-        index c = Aux::Random::integer(k-1);
-        zeta.addToSubset(c, v);
-    });
-
-    assert (zeta.numberOfSubsets() == k);
-
-    G.forNodePairs([&](node u, node v){
-        if (zeta.subsetOf(u) == zeta.subsetOf(v)) {
-            if (Aux::Random::probability() <= pin) {
-                G.addEdge(u, v);
-            }
-        } else {
-            if (Aux::Random::probability() <= pout) {
-                G.addEdge(u, v);
-            }
+    // Sequence of all nodes grouped by cluster. Here the vertices of each cluster are stored
+    // sequentially one after the other
+    std::vector<node> clustersSequence;
+
+    // Index in `clustersSequence` where each cluster ends. E.g., clusterEnd[i] is the index in
+    // `clustersSequence` of the first vertex in the (i + 1)-th cluster.
+    std::vector<size_t> clusterEnd(k);
+
+    {
+        // To assign vertices to clusters uniformly at random
+        std::uniform_int_distribution<index> clusterDistr{0, k - 1};
+
+        // Nodes grouped by cluster
+        std::vector<std::vector<node>> clusters(k);
+
+        // Assign nodes to clusters
+        graphBuilder.parallelForNodes([&zeta = zeta, &clusters, &clusterDistr](node u) {
+            const index cluster = clusterDistr(Aux::Random::getURNG());
+            zeta.addToSubset(cluster, u);
+#pragma omp critical
+            clusters[cluster].push_back(u);
+        });
```

Comment:
Entering a parallel region just to do an immediate `#pragma omp critical` is unlikely to be beneficial. You could use an array of per-cluster locks instead.